core/src/main/java/org/apache/accumulo/fate/zookeeper/DistributedReadWriteLock.java

```diff
@@ -218,22 +228,69 @@ public boolean tryLock() {
       }
       SortedMap<Long,byte[]> entries = qlock.getEarlierEntries(entry);
       Iterator<Entry<Long,byte[]>> iterator = entries.entrySet().iterator();
-      if (!iterator.hasNext())
+      if (!iterator.hasNext()) {
         throw new IllegalStateException("Did not find our own lock in the queue: " + this.entry
             + " userData " + new String(this.userData, UTF_8) + " lockType " + lockType());
-      return iterator.next().getKey().equals(entry);
+      }
+      if (iterator.next().getKey().equals(entry)) {
+        return true;
+      }
+      if (failBlockers) {
+        // Loop through all of the prior transactions that are waiting to
+        // acquire this write lock. If the transaction has not succeeded or failed,
+        // then fail it and return false from this method so that Utils.reserveX()
+        // will call this method again and will re-check the prior transactions.
+        boolean result = true;
+        while (iterator.hasNext()) {
+          Entry<Long,byte[]> e = iterator.next();
+          Long txid = e.getKey();
+          if (!txid.equals(entry)) {
+            if (store.isReserved(txid)) {
+              result &= false;
+            } else {
+              store.reserve(txid);
```

Comment:
The check for reserved and then reserving is a racy.  Could add a new method that tries to reserve atomically.

```suggestion
            if (store.tryReserve(txid)) {

```

Reply:
Addressed in 2b8fc42ebd19b8c9b3861ab9d5abaa3d34a6888f